Create scalahost sbt plugin, fixes #673.

olafurpg opened the pull request:
- Injects scalahost compiler plugin for 2.11/2.12 projects
- Ignores 2.10 projects
- Defaults to build version.value, but possible to override with -Dscalahost.version
- Possible to use local assembly jar for fast iteration speeds during development.

xeno-by commented on build.sbt:
```diff
@@ -282,8 +321,6 @@ lazy val readme = scalatex.ScalatexReadme(
 lazy val sharedSettings = Def.settings(
   scalaVersion := ScalaVersion,
-  crossScalaVersions := ScalaVersions,
-  crossVersion := CrossVersion.binary,
```
How about we keep these lines in place and just override these settings in scalahostSbt? It looks like this will significantly reduce the amount of required changes.

olafurpg replied:
Good idea, I started this change before I realised it was only two setting keys.

xeno-by commented on build.sbt:
```diff
@@ -21,6 +22,19 @@ lazy val scalametaRoot = Project(
   sharedSettings,
   unidocSettings,
+  // needs to be a command since scripted requires publishing across scala-versions
+  commands += Command.command("scripted") { state =>
```
What does "scripted" mean?

xeno-by commented on build.sbt:
```diff
@@ -21,6 +22,19 @@ lazy val scalametaRoot = Project(
+    "very scalahostSbt/scripted" ::
+    state
+  },
+  commands += Command.command("testAll") { state =>
```
How about we rename `testAll` to something different, e.g. `ci`, so it doesn't make an impression of something that just runs all tests aggregated across projects?

olafurpg replied:
Good idea, done.

xeno-by commented on build.sbt:
```diff
@@ -201,11 +216,35 @@ lazy val scalahost = Project(
 
+lazy val scalahostSbt = Project(
+  id   = "scalahostSbt",
+  base = file("sbt-plugins/scalahost")
```
Why not have it under `scalahostSbt` like you do in scalafmt and scalafix?

olafurpg replied:
You mean scalahost-sbt? That's fine too, I was unsure where to put it.

xeno-by commented on sbt-plugins/scalahost/src/sbt-test/sbt-scalahost/helloworld/test:
```diff
@@ -0,0 +1,5 @@
+# compile should not affect scalafix
+> dumpsemanticdb
+# TODO(olafur) verify that 2.11 and 2.12 got dumped
```
Do you want to verify this in this PR, or plan to submit a follow-up?

olafurpg replied:
This is a bug, see #650. I added a link to the issue in the test.

olafurpg replied:
I prefer not to block this PR by this bug. My personal use-case for this sbt plugin is to use #666, which I hope to bundle into a PR tomorrow.

xeno-by commented on sbt-plugins/scalahost/src/main/scala/scala/meta/scalahost/sbt/ScalahostSbtPlugin.scala:
```diff
@@ -0,0 +1,41 @@
+  // NOTE. Only possible with one scalaVersion, not cross-build projects.
+  val scalahostPackageJar: Option[String] = sys.props.get("scalahost.packagejar")
+  override def projectSettings = Seq(
+    commands += Command.command("dumpsemanticdb")(s => "test:compile" :: s),
```
Why do we need a separate command for this if simply running `compile` should be enough?

olafurpg replied:
Before this PR, I only ran test:compile for >2.11 scala version projects and printed out useful info. Since I don't do that anymore, I've remove this command.

xeno-by commented on build.sbt:
```diff
@@ -325,6 +365,8 @@ lazy val mergeSettings = Def.settings(
 
 def computePreReleaseVersion(LibrarySeries: String): String = {
+  // useful to provide custom version number when running sbt scripted tests, to avoid republishing 2.11/2.12
+  if (sys.props.contains("scalameta.version")) return sys.props("scalameta.version")
```
Shouldn't this then go in `LibraryVersion` at the top of the file?

xeno-by commented on project/project/plugins.sbt:
```diff
@@ -0,0 +1 @@
+addSbtPlugin("io.get-coursier" % "sbt-coursier" % "1.0.0-M15")
```
Nice!

xeno-by commented on scalahost-sbt/src/sbt-test/sbt-scalahost/helloworld/build.sbt:
```diff
@@ -0,0 +1,19 @@
+TaskKey[Unit]("check") := {
+  val contents = IO.read(file("semanticdb"))
+  assert(!contents.contains("p1"))
+  // TODO(olafur) enable once https://github.com/scalameta/scalameta/pull/686 is merged.
```
Done

olafurpg commented:
I updated the sbt plugin to automatically upgrade old patch scalaVersion, fixing #681. This means that for example 2.12.0 projects will use scalahost:2.12.1. 

I also replaced the `compilerPlugin` "helper" in sbt (it seems broken for 2.12, picks the javadoc jar) with a custom `Xplugin:/scalahost.jar`.

xeno-by commented on scalahost-sbt/src/main/scala/scala/meta/scalahost/sbt/ScalahostSbtPlugin.scala:
```diff
@@ -0,0 +1,66 @@
+                   |Please report the output below at https://github.com/scalameta/scalameta/issues
+                   |Scala version: ${scalaVersion.value}
+                   |Cross version: ${crossScalaVersions.value}
+                   |Jar regex: $jarRegex
```
When can this happen?

olafurpg replied:
When the update report does not contain scalahost.jar in an expected format.

xeno-by commented:
LGTM

olafurpg commented:
BTW. The CI is super slow because it's spending like 10 minutes resolving dependencies. It seems tha the drone cache plugin is disabled when .drone.yml changes (wtf? 😭 )

olafurpg commented:
I ran this sbt plugin through the community build and most critical errors seems to have been fixed now. https://github.com/scalameta/scalameta/issues/665 appears in a few places. I was able to collect ~2.2 million symbols with the latest scalahost version (v1.6.0-687) compared to ~210k symbols last week.
